### ISSUE TITLE
Linking ABI: import the memory and table?

### ISSUE THREAD

**Issue by sunfishcode** [WebAssembly/tool-conventions#23]
Currently, clang emits code that imports the stack pointer, but doesn't import the `memory` or `table`. It doesn't export the ones it defines either. Clearly the linker can still do what it needs, but it would be nice to have some consistency here. Is it best to import the `memory` and `table`, or do something else?

**Comment by dschuff** [comment]
I guess if we want to maintain the "object files are runnable" idea then we probably need to locally define the stack pointer (along with memory and table) and make the linker more magical. Otherwise it probably makes sense to have object files just import everything.
Even in that case (i.e. the case where we give up on saying object files are runnable) it probably makes sense to have some consistent idea of how valid or how close to runnable object files are? E.g. we could at least say "object files are valid wasm" (does that currently hold?). If we do that then we'd want to locally define memory and table (and make it shared if needed, etc) to keep that property.

**Comment by sbc100** [comment]
We do already say that they should be valid wasm:  "These modules are not expected to be directly executable and have certain constraints on them, but are otherwise well-formed WebAssembly modules".  

 I don't have strong opinions about whether llvm-generated object should import or declare their memories and tables, either one seems reasonable, and does not make the more or less runnable does it?

**Comment by sunfishcode** [comment]
I agree. I'm now less interested in making .o files directly runnable than I initially was. But I am still interested in consistency where it makes sense, and this seems like an area where it makes sense.

**Comment by guybedford** [comment]
I haven't followed the most recent updates on the web assembly stack pointer conventions, so this may be completely off track, but I do think it's worth considering (or ruling out, and if not already) that a nice property to have might be the ability to merge multiple web assembly binaries into a single binary (a sort of composibility optimization for better network and parse times in the final production module graph, like we have with JS modules).

For example say a trend is to publish small wasm + JS packages to npm and we want Webpack users to be able to bundle that up for production. Usually when we have lots of separate JS files we want to combine them into common JS files for production, and ES modules have this composibility property for us. If the common wasm binaries could be squashed into a single common wasm binary that might be a nice optimization to have as well. This arbitrary composibility just kind of appeals as a concept.

As well as a stack pointer it seems like this would mean having a memory data start offset pointer from the base memory position so the combined memory could be layed out as `data1|stack1|data2|stack2|etc` (yes there are cross-memory corruption concerns here, but for a production build as an optimization it would be debuggable in the development "unoptimized" form and is no different than static linking memory corruption potential). I don't believe such a convention would cause issues for memory grow operations as this is always deemed new managed heap space above the stack anyway? New custom global names could be created to set these offsets in combined binary.

So if data memory operations could be understood as data-memory-global-relative in the runnable binary format via some runtime convention, that could might enable these sorts of optimizations.

Now this might just be a pipe dream, but having such a memory convention work in runtime, while also supporting a memory position data offsets that could easily be transformed and relinked for composition in other modules, such a convention may then make sense as the object linking convention itself then if it can be achieved.

I just thought this naively worth bringing here, but I understand the practicalities are another thing entirely, and there are may other cross cutting concerns for the ABI that I am likely entirely unaware of as well!

**Comment by sbc100** [comment]
I might be wrong but it sounds like you are talking about dynamic linking here?  Have you see the current proposal on that: https://github.com/WebAssembly/tool-conventions/blob/master/DynamicLinking.md

**Comment by guybedford** [comment]
I was actually referring to static linking optimisations (merging) for wasm
executables, but thanks for the link I will have a look.

**Comment by guybedford** [comment]
@sbc100 the principles of dynamic linking are definitely very similar though. Is there a reason a global couldn't be used for both the `tablesize` and `memoryBase` values? Let me know if there is a better place to discuss this too. I understand these globals would need to be incremented if imported for other linkages, but may not work out timing-wise to do in a start function. I guess I'm kind of trying to work out if the dynamic linking ABI could itself become a standard library ABI form for sharing code on npm etc, even for static linking.

**Comment by sunfishcode** [comment]
This is implemented in LLVM in r319989.
